Zero initReg in genSetGSSecurityCookie

echesakovMSFT opened the pull request:
The issue in #24278:
1. register **initReg** is used for initializing GSSecurityCookie in prolog (in the example below **initReg** is **r4**);
2. the same register is live and holds enregistered must-init variable in a basic block next to prolog.
```
; Final local variable assignments
;
;  V00 loc0         [V00,T01] (  2,  1.50)     ref  ->   r4         must-init class-hnd exact
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   "OutgoingArgSpace"
;  V02 tmp1         [V02    ] (  1,  1   )     blk (32) [sp+0x04]   do-not-enreg[XF] must-init addr-exposed unsafe-buffer "stackallocLocal"
;  V03 tmp2         [V03,T00] (  2,  2   )     ref  ->   r4         class-hnd exact "NewObj constructor temp"
;* V04 GsCookie     [V04    ] (  0,  0   )     int  ->  zero-ref    do-not-enreg[X] must-init addr-exposed "GSSecurityCookie"

*************** After end code gen, before unwindEmit()
G_M59476_IG01:        ; func=00, offs=000000H, size=0026H, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, nogc <-- Prolog IG

IN001e: 000000  4B      push    {r4,r11,lr}
IN001f: 000004  2B      sub     sp, 44
IN0020: 000006  4B      add     r11, sp, 48
IN0021: 00000A  4B      sub     r1, r11, 44
IN0022: 00000E  2B      movs    r4, 0
IN0023: 000010  2B      movs    r0, 0
IN0024: 000012  2B      stm     r1!, {r0,r4}
IN0025: 000014  2B      stm     r1!, {r0,r4}
IN0026: 000016  2B      stm     r1!, {r0,r4}
IN0027: 000018  2B      stm     r1!, {r0,r4}
IN0028: 00001A  2B      str     r4, [r1]
IN0029: 00001C  4B      movw    r4, 0x5678
IN002a: 000020  4B      movt    r4, 0x1234
IN002b: 000024  2B      str     r4, [sp+0x24]

G_M59476_IG02:        ; offs=000026H, size=0032H, gcrefRegs=0010 {r4}, byrefRegs=0000 {}, byref, isz
```
3. During GC **r4** is dereferenced and segfaults 
```
(lldb) bt
* thread #1, name = 'corerun', stop reason = signal SIGSEGV: invalid address (fault address: 0x12345678)
  * frame #0: 0xb655594c libcoreclr.so`Object::ValidateInner(int, int, int) [inlined] Object::GetGCSafeMethodTable(this=0x12345678) const at object.h:471
    frame #1: 0xb655594c libcoreclr.so`Object::ValidateInner(this=0x12345678, bDeep=YES, bVerifyNextHeader=YES, bVerifySyncBlock=YES) at object.cpp:590
    frame #2: 0xb66b3b20 libcoreclr.so`GcInfoDecoder::ReportRegisterToGC(this=0xbed9db48, regNum=<unavailable>, gcFlags=0, pRD=<unavailable>, flags=0, pCallBack=(libcoreclr.so`GcEnumObject(void*, OBJECTREF*, unsigned int) + 1 at gcenv.ee.common.cpp:148), hCallBack=<unavailable>)(void*, OBJECTREF*, unsigned int), void*) at gcinfodecoder.cpp:1590
    frame #3: 0xb66b4220 libcoreclr.so`GcInfoDecoder::ReportSlotToGC(this=0xbed9db48, slotDecoder=0xbed9da9c, slotIndex=0, pRD=0xbed9e018, reportScratchSlots=false, inputFlags=0, pCallBack=(libcoreclr.so`GcEnumObject(void*, OBJECTREF*, unsigned int) + 1 at gcenv.ee.common.cpp:148), hCallBack=0xbed9e5fc)(void*, OBJECTREF*, unsigned int), void*) at gcinfodecoder.h:659
    frame #4: 0xb66b1ec8 libcoreclr.so`GcInfoDecoder::EnumerateLiveSlots(this=0xbed9db48, pRD=<unavailable>, reportScratchSlots=false, inputFlags=0, pCallBack=<unavailable>, hCallBack=<unavailable>)(void*, OBJECTREF*, unsigned int), void*) at gcinfodecoder.cpp:791
    frame #5: 0xb64eecb2 libcoreclr.so`EECodeManager::EnumGcRefs(this=<unavailable>, pRD=0xbed9e018, pCodeInfo=0xbed9df38, flags=0, pCallBack=(libcoreclr.so`GcEnumObject(void*, OBJECTREF*, unsigned int) + 1 at gcenv.ee.common.cpp:148), hCallBack=<unavailable>, relOffsetOverride=<unavailable>)(void*, OBJECTREF*, unsigned int), void*, unsigned int) at eetwain.cpp:5121
    frame #6: 0xb65f16d0 libcoreclr.so`GcStackCrawlCallBack(pCF=0xbed9dd18, pData=0xbed9e5fc) at gcenv.ee.common.cpp:290
    frame #7: 0xb656d016 libcoreclr.so`Thread::MakeStackwalkerCallback(this=0x00063640, pCF=0xbed9dd18, pCallback=(libcoreclr.so`GcStackCrawlCallBack(CrawlFrame*, void*) + 1 at gcenv.ee.common.cpp:201), pData=0xbed9e5fc, uFramesProcessed=4)(CrawlFrame*, void*), void*, unsigned int) at stackwalk.cpp:870
    frame #8: 0xb656d248 libcoreclr.so`Thread::StackWalkFramesEx(this=0x00063640, pRD=0xbed9e018, pCallback=(libcoreclr.so`GcStackCrawlCallBack(CrawlFrame*, void*) + 1 at gcenv.ee.common.cpp:201), pData=0xbed9e5fc, flags=<unavailable>, pStartFrame=0x00000000)(CrawlFrame*, void*), void*, unsigned int, Frame*) at stackwalk.cpp:950
    frame #9: 0xb656dad2 libcoreclr.so`Thread::StackWalkFrames(this=0x00063640, pCallback=(libcoreclr.so`GcStackCrawlCallBack(CrawlFrame*, void*) + 1 at gcenv.ee.common.cpp:201), pData=0xbed9e5fc, flags=34048, pStartFrame=<unavailable>)(CrawlFrame*, void*), void*, unsigned int, Frame*) at stackwalk.cpp:1033
    frame #10: 0xb67f35bc libcoreclr.so`ScanStackRoots(pThread=0x00063640, fn=(libcoreclr.so`WKS::GCHeap::Promote(Object**, ScanContext*, unsigned int) + 1 at gc.cpp:34478), sc=0xbed9e698)(Object**, ScanContext*, unsigned int), ScanContext*) at gcenv.ee.cpp:148
    frame #11: 0xb67f3356 libcoreclr.so`GCToEEInterface::GcScanRoots(fn=(libcoreclr.so`WKS::GCHeap::Promote(Object**, ScanContext*, unsigned int) + 1 at gc.cpp:34478), condemned=<unavailable>, max_gen=<unavailable>, sc=0xbed9e698)(Object**, ScanContext*, unsigned int), int, int, ScanContext*) at gcenv.ee.cpp:177
    frame #12: 0xb67d2196 libcoreclr.so`WKS::gc_heap::mark_phase(condemned_gen_number=<unavailable>, mark_only_p=NO) at gc.cpp:20045
    frame #13: 0xb67cf3ba libcoreclr.so`WKS::gc_heap::gc1() at gc.cpp:15691
    frame #14: 0xb67d89f8 libcoreclr.so`WKS::gc_heap::garbage_collect(n=2) at gc.cpp:0
    frame #15: 0xb67cab88 libcoreclr.so`WKS::GCHeap::GarbageCollectGeneration(this=<unavailable>, gen=<unavailable>, reason=<unavailable>) at gc.cpp:35906
    frame #16: 0xb67ee2da libcoreclr.so`WKS::GCHeap::GarbageCollect(int, bool, int) [inlined] WKS::GCHeap::GarbageCollectTry(generation=<unavailable>) at gc.cpp:35398
    frame #17: 0xb67ee2d2 libcoreclr.so`WKS::GCHeap::GarbageCollect(this=<unavailable>, generation=<unavailable>, low_memory_p=<unavailable>, mode=<unavailable>) at gc.cpp:35332
    frame #18: 0xb65cbd6c libcoreclr.so`GCInterface::Collect(generation=-1, mode=2) at comutilnative.cpp:1185
```


In this situation **r4** must be zeroed at the end of prolog before it becomes live.

Fixes #24278

echesakovMSFT commented:
@BruceForstall @dotnet/jit-contrib @dotnet/arm32-contrib PTAL

BruceForstall commented on src/jit/codegencommon.cpp:
```diff
@@ -6507,6 +6507,8 @@ void CodeGen::genSetGSSecurityCookie(regNumber initReg, bool* pInitRegZeroed)
         instGen_Store_Imm_Into_Lcl(TYP_I_IMPL, EA_PTRSIZE, compiler->gsGlobalSecurityCookieVal,
                                    compiler->lvaGSSecurityCookie, 0, initReg);
+
+        *pInitRegZeroed = false;
```
Unfortunately, this might cause x86 to zero `initReg` later when it doesn't need to, since `instGen_Store_Imm_Into_Lcl` for x86 doesn't touch `initReg`.

echesakovMSFT replied:
Yes, you are right. And it's very hard to notice since we passing `initReg` to `instGen_Store_Imm_Into_Lcl` in x86 case (where it's going to be ignored)

echesakovMSFT replied:
`instGen_Store_Imm_Into_Lcl` was not that useful function - 3 usages in total and its interface is misleading - having a parameter that is not used on a particular architecture. I decided to remove this function completely.

erozenfeld commented on src/jit/codegencommon.cpp:
```diff
@@ -6507,6 +6507,8 @@ void CodeGen::genSetGSSecurityCookie(regNumber initReg, bool* pInitRegZeroed)
         instGen_Store_Imm_Into_Lcl(TYP_I_IMPL, EA_PTRSIZE, compiler->gsGlobalSecurityCookieVal,
                                    compiler->lvaGSSecurityCookie, 0, initReg);
+
+        *pInitRegZeroed = false;
```
We now have `*pInitRegZeroed = false` both in the `if` and the `else`. Can we set it once outside of the if/else?

BruceForstall replied:
It's not set for `_TARGET_XARCH_` in the `if`. Also, it should not be set for `_TARGET_XARCH_` in the `else`.

CarolEidt replied:
This code is a bit difficult to reason about. Is it possible that `initReg` can be `REG_EAX` on `_TARGET_XARCH_`? If so, would we then need to set `*pInitRegZeroed = false`?
It would be great to add more comments to clarify - potentially in a new function header?

echesakovMSFT replied:
This function is indeed difficult to reason and it has one issue that I am fixing in this PR and the conservative assignment `*pInitRegZeroed = false` that Bruce pointed out.

I am looking into splitting the function (or at least if-else block) into `_TARGET_ARMARCH_` and `_TARGET_XARCH_` and see if this helps to improve readability. 

Although, my original plan was to deliver correctness fix in one PR and follow-up later with the refactoring changes now I am thinking about doing both as a part of one PR but in separate commits.

echesakovMSFT commented:
/azp run coreclr-ci

echesakovMSFT commented:
@dotnet/jit-contrib I think this is ready for another round of review:

Here is a summary of what was done:
1. I split **genSetGSSecurityCookie** into two x64/x86 version and arm32/arm64 one.
2. I removed **instGen_Store_Imm_Into_Lcl** which was only used in three places and also had confusing interface - it was accepting `regNumber regToUse` argument and never using the registrer on x86 platform.
3. I added small optimization in **genSetGSSecurityCookie** on x64 - when encoding allows jit will use emitIns_S_I without touching the intermediate register (and setting *pInitRegZeroed to false).

I have already collected superpmi diffs on Linux/arm32 and Linux/arm64 - no diffs. 
I am collecting x64/x86 on Windows and Linux right now.

I also will do the same for NgenDisasm for all the platforms.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -482,6 +482,45 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+        return;
+    }
+
+    noway_assert(compiler->gsGlobalSecurityCookieAddr || compiler->gsGlobalSecurityCookieVal);
```
I realize that this is copied from the pre-existing method, but these should be testing against `nullptr` - or, you could just `noway_assert` that `compiler->gsGlobalSecurityCookieVal` is non-null in the else clause.

echesakovMSFT replied:
@CarolEidt I think you meant add this assertion in the if clause?

CarolEidt replied:
Yes, that's correct. Sorry for the confusion.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -482,6 +482,45 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+// Arguments:
+//     initReg        - register to use as scratch register
+//     pInitRegZeroed - OUT parameter. *pInitRegZeroed set to 'false' if 'initReg' is
+//                      not zero after this call.
```
It seems like the semantics are/should be that it sets `*pInitRegZeroed` to false if and only if it sets `initReg` to a non-zero value.

echesakovMSFT replied:
I have corrected the comments

CarolEidt commented on src/jit/codegenxarch.cpp:
```diff
@@ -48,6 +48,57 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+        //  mov   dword ptr [frame.GSSecurityCookie], eax
+        getEmitter()->emitIns_R_AI(INS_mov, EA_PTR_DSP_RELOC, REG_EAX, (ssize_t)compiler->gsGlobalSecurityCookieAddr);
+        regSet.verifyRegUsed(REG_EAX);
+        getEmitter()->emitIns_S_R(ins_Store(TYP_I_IMPL), EA_PTRSIZE, REG_EAX, compiler->lvaGSSecurityCookie, 0);
```
Shouldn't this path set `*pInitRegZeroed = false`?

echesakovMSFT replied:
Added the check for initReg being EAX and updating *pInitRegZeroed

echesakovMSFT commented:
No diffs with SuperPMI on Windows with x86 clrjit.dll and arm/arm64 protononjit.dll.
Expected diffs (operand register mismatch) with SuperPMI on Windows with x64 clrjit.dll due to the change in 91f2f18bdaa29844c5b899d0c710cab69e8af879

No diffs with CrossGen on Windows with x86/x64.
Still running CrossGen on Windows with arm/arm64 but expect no diffs there as well

echesakovMSFT commented:
@dotnet/jit-contrib Can you please take another look?

BruceForstall commented on tests/src/JIT/Regression/JitBlue/GitHub_24278/GitHub_24278.ilproj:
```diff
@@ -0,0 +1,25 @@
+    <ProjectGuid>{95DFC527-4DC1-495E-97D7-E94EE1F7140D}</ProjectGuid>
+    <OutputType>Exe</OutputType>
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
+    <CLRTestPriority>0</CLRTestPriority>
```
nit: I don't think you need to set `CLRTestPriority` if it's zero.

erozenfeld commented on src/jit/codegenxarch.cpp:
```diff
@@ -49,6 +49,61 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+//
+// Arguments:
+//     initReg        - register to use as scratch register
+//     pInitRegZeroed - OUT parameter. *pInitRegZeroed set to 'false' if and only if
```
set --> is set

erozenfeld commented on src/jit/codegenarmarch.cpp:
```diff
@@ -567,6 +567,44 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+//
+// Arguments:
+//     initReg        - register to use as scratch register
+//     pInitRegZeroed - OUT parameter. *pInitRegZeroed set to 'false' if and only if
```
set --> is set

erozenfeld replied:
This method is not zero-ing initReg so it sounds strange to me that an out-parameter is named pInitRegZeroed. It also doesn't know or check that it holds zero on input. IMO, it would be clearer to name the argument pInitRegUsed and set it to true if initReg is used for setting the cookie.

CarolEidt replied:
I agree that it's confusing. I might rename it `pInitRegModified`. Theoretically it wouldn't have to set it to false if it had actually zero'd the register, but that doesn't actually happen anywhere. It might be worth a comment here in the header to note that the caller is passing in the `initReg` that it uses in the prolog and therefore it needs to know whether it has been set to a non-zero value (or something to that effect).

BruceForstall replied:
While I don't disagree about it being potentially confusing, note that `pInitRegZeroed` is the name used in many functions that do the same thing.

echesakovMSFT replied:
Created issue https://github.com/dotnet/coreclr/issues/24985 to address this work

erozenfeld commented on src/jit/codegenarmarch.cpp:
```diff
@@ -567,6 +567,44 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+// genSetGSSecurityCookie: Set the "GS" security cookie in the prolog.
+//
+// Arguments:
+//     initReg        - register to use as scratch register
```
as scratch register --> as a scratch register

erozenfeld commented on src/jit/codegenxarch.cpp:
```diff
@@ -49,6 +49,61 @@ void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type, insFla
+// genSetGSSecurityCookie: Set the "GS" security cookie in the prolog.
+//
+// Arguments:
+//     initReg        - register to use as scratch register
```
as scratch --> as a scratch

erozenfeld commented on tests/src/JIT/Regression/JitBlue/GitHub_24278/GitHub_24278.il:
```diff
@@ -0,0 +1,54 @@
+    IL_0005:  ret
+  } // end of method Program::Use
+
+  .method private hidebysig static void  MainMethod() cil managed noinlining
```
It would be good to add a comment to the test to describe what it's doing.

echesakovMSFT commented:
@erozenfeld @CarolEidt @BruceForstall I agree with your suggestions regarding the parameter name changes but I would like to merge this PR as is to address the product issue. Then I can follow-up in another PR with all the refactoring changes
